### PR TITLE
Specification for replay protected signatures

### PR DESCRIPTION
**Pull request by deadalnix**
This follow discussion about using BIP143 hashing scheme for replay protected signatures. This spec is essentially the same as BIP143 with the following detail changed:
 - The way script computation is specified has been generalized to non segwit transactions.
 - nHashType is now tweaked using a 24 bits fork id in order to provide replay protection.

### CODE REVIEW

File: BUIP-HF/replay-protected-sighash.md [hunk 1]
```diff
@@ -0,0 +1,191 @@
+# BUIP-HF Digest for replay protected signature verification accross hard forks
+
+## Abstract
+
+This document describes proposed requirements and design for a reusable signing mechanism ensuring replay protection in the event of a hard fork. It provides a way for users to create transactions which are only valid on one branch of the fork.
+
+The proposed digest algorithm is similar to [BIP143](https://github.com/bitcoin/bips/blob/master/bip-0143.mediawiki) as it is recognized as a superior solution to the currently used algorithm [1] and is already implemented in a wide variety of applications.
+
+The proposed digest algorithm is used when the `SIGHASH_FORKID` bit is set in the signature's sighash type. The way signatures which do not set this is bit are verified is not affected.
```
Comment (ftrader) [inline BUIP-HF/replay-protected-sighash.md]: Suggest rewording due to confusing sentence structure, from 

> The way signatures which do not set this is bit are verified is not affected.

to

> The verification of signatures which do not set this is bit is not affected.

File: BUIP-HF/replay-protected-sighash.md [hunk 2]
```diff
@@ -0,0 +1,191 @@
+# BUIP-HF Digest for replay protected signature verification accross hard forks
+
+## Abstract
+
+This document describes proposed requirements and design for a reusable signing mechanism ensuring replay protection in the event of a hard fork. It provides a way for users to create transactions which are only valid on one branch of the fork.
+
+The proposed digest algorithm is similar to [BIP143](https://github.com/bitcoin/bips/blob/master/bip-0143.mediawiki) as it is recognized as a superior solution to the currently used algorithm [1] and is already implemented in a wide variety of applications.
```
Comment (ftrader) [inline BUIP-HF/replay-protected-sighash.md]: I feel it would be nice to add in brief summary of the benefits here too, notably this part from "Abstract" of BIP143. 

> minimize redundant data hashing in verification, and to cover the input value by the signature. 

That really makes some key benefits clear whereas the linked Motivation section is complex...

File: BUIP-HF/replay-protected-sighash.md [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+* If the sighash type is neither `SINGLE` nor `NONE`, `hashOutputs` is the double SHA256 of the serialization of all output amount (8-byte little endian) with `scriptPubKey` (serialized as scripts inside CTxOuts);
+* If sighash type is `SINGLE` and the input index is smaller than the number of outputs, `hashOutputs` is the double SHA256 of the output amount with `scriptPubKey` of the same index as the input;
+* Otherwise, `hashOutputs` is a `uint256` of `0x0000......0000`.
+
+Notes:
+1. In the [original algorithm](https://en.bitcoin.it/wiki/OP_CHECKSIG), a `uint256` of `0x0000......0001` is committed if the input index for a `SINGLE` signature is greater than or equal to the number of outputs. In this BIP a `0x0000......0000` is committed, without changing the semantics.
+
+#### sighash type
+
+The sighash type is altered to include a 24bits fork id in its most significant bits.
```
Comment (ftrader) [inline BUIP-HF/replay-protected-sighash.md]: 1. suggest to put "fork id" in quotes as it is a special term
2. convention: change "24bits" to  "24-bit" or "24 bit"

File: BUIP-HF/replay-protected-sighash.md [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+
+````cpp
+  ss << ((GetForkID() << 8) | nHashType);
+````
+
+This ensure that the proposed digest algorithm will generate different results on forks using different forkids.
+
+## Implementation
```
Comment (ftrader) [inline BUIP-HF/replay-protected-sighash.md]: I must leave the detailed review ofthe 'Implementation' section to subject experts.

In case the midstate computation code is derived from existing implementation, I would suggest to add a reference to the original code location (since that is not in the BIP143 spec).

Comment (deadalnix) [inline BUIP-HF/replay-protected-sighash.md]: It is in the BIP143 spec, they just inlined them.

**Comment by ftrader** [comment]
What about adding a Copyright (CC0 or otherwise Public Domain) ?

File: BUIP-HF/replay-protected-sighash.md (hunk 4 above)
Comment (ftrader) [inline BUIP-HF/replay-protected-sighash.md]: You're right, now I see it.

**Comment by deadalnix** [comment]
Improved as per comments.

**Comment by ftrader** [comment]
I'm ok with it as-is.

**Comment by gandrewstone** [comment]
looks great!👍

File: BUIP-HF/replay-protected-sighash.md [hunk 5]
```diff
@@ -0,0 +1,191 @@
+# BUIP-HF Digest for replay protected signature verification accross hard forks
+
+## Abstract
+
+This document describes proposed requirements and design for a reusable signing mechanism ensuring replay protection in the event of a hard fork. It provides a way for users to create transactions which are only valid on one branch of the fork.
```
Comment (dgenr8) [inline BUIP-HF/replay-protected-sighash.md]: <strike>only valid on one branch of the fork</strike>
invalid on forks lacking support for the mechanism and a fork-specific ID

File: BUIP-HF/replay-protected-sighash.md [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+
+The proposed digest algorithm is similar to [BIP143](https://github.com/bitcoin/bips/blob/master/bip-0143.mediawiki) as it minimizes redundant data hashing in verification, covers the input value by the signature [1] and is already implemented in a wide variety of applications.
+
+The proposed digest algorithm is used when the `SIGHASH_FORKID` bit is set in the signature's sighash type. The verification of signatures which do not set this is bit is not affected.
+
+## Specification
+
+### Activation
+
+The proposed digest algorithm is only used when the `SIGHASH_FORKID` bit in the signature sighash's type is set. It is defined as follow:
```
Comment (dgenr8) [inline BUIP-HF/replay-protected-sighash.md]: <strike>follow</strike>
follows

File: BUIP-HF/replay-protected-sighash.md [hunk 7]
```diff
@@ -0,0 +1,191 @@
+# BUIP-HF Digest for replay protected signature verification accross hard forks
+
+## Abstract
+
+This document describes proposed requirements and design for a reusable signing mechanism ensuring replay protection in the event of a hard fork. It provides a way for users to create transactions which are only valid on one branch of the fork.
+
+The proposed digest algorithm is similar to [BIP143](https://github.com/bitcoin/bips/blob/master/bip-0143.mediawiki) as it minimizes redundant data hashing in verification, covers the input value by the signature [1] and is already implemented in a wide variety of applications.
```
Comment (dgenr8) [inline BUIP-HF/replay-protected-sighash.md]: <strike>similar to</strike>
adapted from

File: BUIP-HF/replay-protected-sighash.md [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+#### hashSequence
+
+* If none of the `ANYONECANPAY`, `SINGLE`, `NONE` sighash type is set, `hashSequence` is the double SHA256 of the serialization of `nSequence` of all inputs;
+* Otherwise, `hashSequence` is a `uint256` of `0x0000......0000`.
+
+#### scriptCode
+
+In this section, we call `script` the script being currently executed. This means `redeemScript` in case of P2SH, or the `scriptPubKey` in the general case.
+
+* If the `script` does not contain any `OP_CODESEPARATOR`, the `script` is the `script` serialized as scripts inside `CTxOut`.
```
Comment (dgenr8) [inline BUIP-HF/replay-protected-sighash.md]: the `scriptCode` is the `script`

File: BUIP-HF/replay-protected-sighash.md [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+* If the `script` contains any `OP_CODESEPARATOR`, the `scriptCode` is the `script` but removing everything up to and including the last executed `OP_CODESEPARATOR` before the signature checking opcode being executed, serialized as scripts inside `CTxOut`.
+
+Notes:
+1. Contrary to the original algorithm, this one do not use `FindAndDelete` to remove the signature from the script.
```
Comment (dgenr8) [inline BUIP-HF/replay-protected-sighash.md]: does not use

File: BUIP-HF/replay-protected-sighash.md [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+2. Because of 1. it is not possible to create a valid signature within `redeemScript` or `scriptPubkey` as the signature would be part of the digest. This enforce that the signature is in `sigScript` .
```
Comment (dgenr8) [inline BUIP-HF/replay-protected-sighash.md]: Because of 1, it is not
...
enforces

File: BUIP-HF/replay-protected-sighash.md [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+3. In case an opcode require signature checking is present in `sigScript`, `script` is effectively `sigScript`. However, for reason similar to 2. , it is not possible to provide a valid signature in that case.
```
Comment (dgenr8) [inline BUIP-HF/replay-protected-sighash.md]: that requires

File: BUIP-HF/replay-protected-sighash.md (hunk 8 above)
Comment (deadalnix) [inline BUIP-HF/replay-protected-sighash.md]: Not in present of OP_CODESEPARATOR

Comment (deadalnix) [inline BUIP-HF/replay-protected-sighash.md]: Ho I got it.

**Comment by deadalnix** [comment]
Made changes suggested by @dgenr8 .
